types/redux-form/lib/reduxForm.d.ts

```diff
@@ -1,571 +1,140 @@
-     * of invalid.
-     */
     valid: boolean;
+    warning: any;
+}
 
-    /**
-     * A generic warning for the entire form given by the `_warning` key in the result from the
-     * synchronous warning function.
-     */
-    warning: string;
+export interface ConfigProps<FormData = {}, P = {}> {
+    form: string;
+    asyncBlurFields?: string[];
+    asyncValidate?(values: FormData, dispatch: Dispatch<any>, props: P & InjectedFormProps<FormData, P>, blurredField: string): Promise<any>;
+    destroyOnUnmount?: boolean;
+    enableReinitialize?: boolean;
+    forceUnregisterOnUnmount?: boolean;
+    getFormState?(state: any): FormStateMap;
+    immutableProps?: string[];
+    initialValues?: Partial<FormData>;
+    keepDirtyOnReinitialize?: boolean;
+    onChange?: (values: Partial<FormData>, dispatch: Dispatch<any>, props: P & InjectedFormProps<FormData, P>) => void;
+    onSubmit?: FormSubmitHandler<FormData, P & InjectedFormProps<FormData, P>>;
+    onSubmitFail?(errors: FormErrors<FormData>, dispatch: Dispatch<any>, submitError: any, props: P & InjectedFormProps<FormData, P>): void;
+    onSubmitSuccess?(result: any, dispatch: Dispatch<any>, props: P & InjectedFormProps<FormData, P>): void;
+    propNamespace?: string;
+    pure?: boolean;
+    shouldValidate?(params: ValidateCallback<FormData, P>): boolean;
+    shouldAsyncValidate?(params: AsyncValidateCallback<FormData>): boolean;
+    touchOnBlur?: boolean;
+    touchOnChange?: boolean;
+    persistentSubmitErrors?: boolean;
+    validate?(values: FormData, props: P & InjectedFormProps<FormData, P>): FormErrors<FormData>;
+    warn?(values: FormData, props: P & InjectedFormProps<FormData, P>): FormWarnings<FormData>;
 }
 
-/**
- * These are the props that will be passed to your form component.
- * Your form component's props can extend this interface.
- */
-export interface FormProps<FormData extends DataShape, P, S> extends Partial<StrictFormProps<FormData, P, S>> {
+export interface FormInstance<FormData, P> extends Component<P> {
+    dirty: boolean;
+    invalid: boolean;
+    pristine: boolean;
+    registeredFields: RegisteredFieldState[];
+    reset(): void;
+    submit(): Promise<any>;
+    valid: boolean;
+    values: Partial<FormData>;
+    wrappedInstance: ReactElement<P & InjectedFormProps<FormData, P>>;
+}
 
+export interface DecoratedComponentClass<FormData, P> {
+    new (props?: P, context?: any): FormInstance<FormData, P>;
 }
+
+export type FormDecorator<FormData, P, Config> =
+    (component: ComponentType<P & InjectedFormProps<FormData, P>>) => DecoratedComponentClass<FormData, P & Config>;
+
+declare function reduxForm<FormData = {}, P = {}>(
+    config: ConfigProps<FormData, P>
+): FormDecorator<FormData, P, Partial<ConfigProps<FormData, P>>>;
+
+declare function reduxForm<FormData = {}, P = {}>(
+    config: Partial<ConfigProps<FormData, P>>
+): FormDecorator<FormData, P, ConfigProps<FormData, P>>;
```

Comment:
@LKay, i see You changed `reduxForm` function definition in this PR ([see original definition of this function](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/18168/files#diff-d9e48aae97c787121e5ed719108dc0b2L12)) by adding `Partial<...>` - it seems to me that here is a copy-paste error because 
1) in this function overload **Partial** Config is taken as input and **non-partial** Config is included in output type (I'm guessing if Partial Config should also be in output)
2) in another overload of this function (few lines above) **non-partial** Config is taken as input and **Partial** Config is included in output type

My first thought is that if you pass partial config then you should get back type with partial config, not with non-partial config (and vice versa). 

Is this a bug or is there smth i don't understand?

----
i'll add an example of a function that produces compile-time error with current types:
```TypeScript
export function withReduxFormCorrectNonPartial2<FormData, P>(component: ComponentType<P & InjectedFormProps<FormData, any>>, config: ConfigProps<FormData, P>)
: DecoratedComponentClass<FormData, P & ConfigProps<FormData, P>> {
	return reduxForm(config)(component);
}
```
```
    TS2322: Type 'DecoratedComponentClass<FormData, P & Partial<ConfigProps<FormData, P>>>' is not assignable to type 'DecoratedComponentClass<FormData, P & ConfigProps<FormData, P>>'.
  Type 'P & Partial<ConfigProps<FormData, P>>' is not assignable to type 'P & ConfigProps<FormData, P>'.
    Type 'P & Partial<ConfigProps<FormData, P>>' is not assignable to type 'ConfigProps<FormData, P>'.
      Property 'form' is optional in type 'P & Partial<ConfigProps<FormData, P>>' but required in type 'ConfigProps<FormData, P>'.
```
The error disappeared after i replaced return types of the overloaded function definitions.

Please let me know, what You think, because i'm new to react and i haven't investigated redux-forms much.